sql/analyzer/apply_indexes_for_subquery_comparisons.go

```diff
@@ -31,52 +72,110 @@ import (
 // 4. The Child is a *plan.ResolvedTable.
 // 5. The referenced field in the Child is indexed.
 func applyIndexesForSubqueryComparisons(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	aliases, err := getTableAliases(n, scope)
-	if err != nil {
-		return nil, transform.SameTree, err
-	}
+	var recurse func(n sql.Node, inScope *aliasScope) (sql.Node, transform.TreeIdentity, error)
+	recurse = func(n sql.Node, inScope *aliasScope) (sql.Node, transform.TreeIdentity, error) {
+		var alias, target string
+		var err error
+		switch n := n.(type) {
+		case *plan.TableAlias:
+			alias = n.Name()
+			switch c := n.Child.(type) {
+			case *plan.ResolvedTable, *plan.SubqueryAlias, *plan.ValueDerivedTable, *plan.TransformedNamedNode, *plan.RecursiveTable, *plan.DeferredAsOfTable:
+				target = c.(sql.Nameable).Name()
+			case *plan.IndexedTableAccess:
+				target = c.Name()
+			case *plan.RecursiveCte:
+			case *plan.UnresolvedTable:
+				panic("Table not resolved")
+			default:
+				panic(fmt.Sprintf("Unexpected child type of TableAlias: %T", c))
+			}
+		default:
+		}
 
-	return transform.Node(n, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
-		switch node := node.(type) {
-		case *plan.Filter:
-			var replacement sql.Node
-			if eq, isEqual := node.Expression.(*expression.Equals); isEqual {
-				replacement = getIndexedInSubqueryFilter(ctx, eq.Left(), eq.Right(), node, true, scope, aliases)
-			} else if is, isInSubquery := node.Expression.(*plan.InSubquery); isInSubquery {
-				replacement = getIndexedInSubqueryFilter(ctx, is.Left, is.Right, node, false, scope, aliases)
+		if alias != "" {
+			err = inScope.add(alias, target)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+		}
+
+		_, isOp := n.(sql.OpaqueNode)
+		outScope := inScope
+		if isOp {
+			outScope = inScope.new()
+		}
+
+		children := n.Children()
+		var newChildren []sql.Node
+		for i, c := range children {
+			child, same, _ := recurse(c, outScope)
+			if !same {
+				if newChildren == nil {
+					newChildren = make([]sql.Node, len(children))
+					copy(newChildren, children)
+				}
+				newChildren[i] = child
```

Comment:
I'm assuming there's not a good way to rewrite this using a `transform` function, but I couldn't tell exactly what the gap was.

Reply:
i wanted an in-order traversal, because i think that is the natural structure of the alias information we are collecting here. we used a bottom-up transform before, i could write a rule that finds the top node of every scope, and then recursively calls the old rule. i'd like to increasingly move towards in-order traversals for anything that relies on inter-scope information.